src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs

```diff
@@ -130,84 +133,102 @@ public bool TryGetValueTask(out ValueTask valueTask, object? keepAlive = null, C
         }
     }
 
-    public Task GetFinalTask() => _finalTaskSource.Task;
+    /// <summary>
+    /// Gets a <see cref="Task"/> that will transition to a completed state with the last transition of this source, i.e. into <see cref="State.Completed"/>.
+    /// </summary>
+    /// <returns>The <see cref="Task"/> that will transition to a completed state with the last transition of this source.</returns>
+    public Task GetFinalTask(object? keepAlive)
+    {
+        lock (this)
+        {
+            return _finalTaskSource.GetTask(keepAlive);
+        }
+    }
 
     private bool TryComplete(Exception? exception, bool final)
     {
+        // Dispose the cancellation registration before completing the task, so that it cannot run after the awaiting method returned.
+        // Dispose must be done outside of lock since it will wait on pending cancellation callbacks that can hold the lock from another thread.
         CancellationTokenRegistration cancellationRegistration = default;
-        try
+        lock (this)
         {
-            lock (this)
+            cancellationRegistration = _cancellationRegistration;
+            _cancellationRegistration = default;
+        }
+        cancellationRegistration.Dispose();
+
+        lock (this)
+        {
+            try
             {
-                try
+                State state = _state;
+
+                // Completed: nothing to do.
+                if (state == State.Completed)
                 {
-                    State state = _state;
+                    return false;
+                }
 
-                    // Completed: nothing to do.
-                    if (state == State.Completed)
-                    {
-                        return false;
-                    }
+                // The task was non-finally completed without having anyone awaiting on it.
+                // In such case, discard the temporary result and replace it with this final completion.
+                if (state == State.Ready && !_hasWaiter && final)
```

Comment:
This allows to ignore `RECEIVE` event notification that no `ReadAsync` is waiting on, allowing `AbortRead` to override them.